src/plugins/embeddable/server/plugin.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreSetup, CoreStart, Plugin, SavedObjectReference } from 'kibana/server';
+import { identity } from 'lodash';
+import {
+  EmbeddableFactoryRegistry,
+  EnhancementsRegistry,
+  EnhancementRegistryDefinition,
+  EnhancementRegistryItem,
+  EmbeddableRegistryDefinition,
+} from './types';
+import {
+  extractBaseEmbeddableInput,
+  injectBaseEmbeddableInput,
+  telemetryBaseEmbeddableInput,
+} from '../common/lib/migrate_base_input';
+import { SerializableState } from '../../kibana_utils/common';
+import { EmbeddableInput } from '../common/types';
+
+export interface EmbeddableSetup {
+  registerEmbeddableFactory: (factory: EmbeddableRegistryDefinition) => void;
+  registerEnhancement: (enhancement: EnhancementRegistryDefinition) => void;
+}
+
+export class EmbeddableServerPlugin implements Plugin<object, object> {
+  private readonly embeddableFactories: EmbeddableFactoryRegistry = new Map();
+  private readonly enhancements: EnhancementsRegistry = new Map();
+
+  public setup(core: CoreSetup) {
+    return {
+      registerEmbeddableFactory: this.registerEmbeddableFactory,
+      registerEnhancement: this.registerEnhancement,
+    };
+  }
+
+  public start(core: CoreStart) {
+    return {
+      telemetry: this.telemetry,
```

Comment:
I guess this is called from the container plugin (e.g. dashboard) to actually build the telemetry which will get added in a separate PR?

If that's the case it seems like this API is not enough to build the "reducer style" api we talked about for collecting telemetry across multiple objects - it looks like `telemetry` should take an optional `telemetryData` which, when set, replaces the `telemetryBaseEmbeddableInput` call.

Reply:
the base embeddable input call still needs to be there, but i agree we need to add an optional telemetryData parameter

Reply:
Yep, that's what I meant - not sure whether typescript allows it but essentially this: ```ts
private telemetry = (state: EmbeddableInput, telemetryData: unknown = telemetryBaseEmbeddableInput(state)) => {
```